fix[DP][v1]: Prevent hangs from mismatched worker configurations

ayushsatyam146 opened the pull request:
Fixes #26187. Previously, vLLM could hang when the leader and workers had different EPLB
configurations (e.g., --enable-eplb-log-balancedness), causing only some
workers to attempt All-Reduce calls.

Changes:
- Extended parallel config hash to include EPLB parameters
- Added config hash exchange during DP coordinator handshake
- Added validation that fails fast on config mismatch with clear error

njhill commented on vllm/v1/engine/core.py:
```diff
@@ -681,17 +681,19 @@ def _perform_handshake(
+                "dp_stats_address": dp_stats_address,
+            }
+            if vllm_config.parallel_config.data_parallel_size > 1:
+                ready_msg["config_hash"] = vllm_config.parallel_config.compute_hash()
```
Suggest to rename this for clarity
```suggestion
                ready_msg["parallel_config_hash"] = vllm_config.parallel_config.compute_hash()
```

njhill commented on vllm/v1/engine/utils.py:
```diff
@@ -867,19 +868,26 @@ def wait_for_engine_startup(
-                            "_data_parallel_master_port_list",
-                            "data_parallel_size",
-                        )
+                        "data_parallel_master_ip": (
+                            parallel_config.data_parallel_master_ip
+                        ),
+                        "data_parallel_master_port": (
+                            parallel_config.data_parallel_master_port
+                        ),
+                        "_data_parallel_master_port_list": (
+                            parallel_config._data_parallel_master_port_list
+                        ),
+                        "data_parallel_size": parallel_config.data_parallel_size,
```
what's the reason for changing this?

njhill commented on vllm/v1/engine/utils.py:
```diff
@@ -73,6 +73,7 @@ class EngineHandshakeMetadata:
 
     addresses: EngineZmqAddresses
     parallel_config: dict[str, Union[int, str, list[int]]]
+    config_hash: Optional[str] = None
```
suggest rename for clarity
```suggestion
    parallel_config_hash: Optional[str] = None
```

njhill reviewed (commented):
Thanks @ayushsatyam146 I think this looks good.

Given that we are already propagating some parameters from the primary worker, we could also consider doing that automatically for some of these too (like the eplb config), so that it doesn't need to be set on the other workers at all.

I guess the only issue there is that it might not be easy to detect and warn about conflicts i.e. when something is set explicitly differently on non-primary worker.

ayushsatyam146 committed:
Filter changed files
fix[DP][v1]: Prevent hangs from mismatched worker configurations
Previously, vLLM could hang when the leader and workers had different EPLB
configurations (e.g., --enable-eplb-log-balancedness), causing only some
workers to attempt All-Reduce calls.

Changes:
- Extended parallel config hash to include EPLB parameters
- Added config hash exchange during DP coordinator handshake
- Added validation that fails fast on config mismatch with clear error

Signed-off-by: Ayush Satyam <ayushsatyam146@gmail.com>

ayushsatyam146 commented:
Hi @njhill thanks for the review. I have addressed all the concerns raised by you. Can you please take a look and review this, Thanks!

njhill reviewed (approved):
Thanks @ayushsatyam146!
